src/utils.py

```diff
@@ -83,45 +96,54 @@ def is_ipv4(ip: str) -> bool:
 
 
 # Service functions
-def _systemctl(action: str, service_name: str) -> NoReturn:
+def _systemctl(action: str, service_name: str) -> None:
     subprocess.run(["systemctl", action, service_name]).check_returncode()
 
 
-def service_start(service_name: str) -> NoReturn:
+def service_start(service_name: str) -> None:
+    """Starts a given service."""
     _systemctl("start", service_name)
 
 
-def service_restart(service_name: str) -> NoReturn:
+def service_restart(service_name: str) -> None:
+    """Restarts a given service."""
     _systemctl("restart", service_name)
 
 
-def service_stop(service_name: str) -> NoReturn:
+def service_stop(service_name: str) -> None:
+    """Stops a given service."""
     _systemctl("stop", service_name)
 
 
-def service_enable(service_name: str) -> NoReturn:
+def service_enable(service_name: str) -> None:
+    """Enables a given service."""
     _systemctl("enable", service_name)
 
 
-def systemctl_daemon_reload():
+def systemctl_daemon_reload() -> None:
+    """Runs `systemctl daemon-reload`."""
     subprocess.run(["systemctl", "daemon-reload"]).check_returncode()
 
 
-def ip_from_default_iface() -> str:
+def ip_from_default_iface() -> Optional[str]:
+    """Returns a Ip address from the default interface."""
     default_gateway = netifaces.gateways()["default"]
     if netifaces.AF_INET in default_gateway:
         _, iface = netifaces.gateways()["default"][netifaces.AF_INET]
         default_interface = netifaces.ifaddresses(iface)
         if netifaces.AF_INET in default_interface:
             return netifaces.ifaddresses(iface)[netifaces.AF_INET][0].get("addr")
+    return None
 
 
-def ip_from_iface(subnet: str) -> str:
+def ip_from_iface(subnet: str) -> Optional[str]:
+    """Returns Ip address from a given subnet."""
     try:
         target_network = IPNetwork(subnet)
         networks = get_local_ipv4_networks()
         for network in networks:
             if network.ip in target_network:
                 return network.ip.format()
+        return None
```

Comment:
Also nitpicky, still, cleaner and shorter?
```suggestion
        return next((network.ip.format() for network in networks if network.ip in target_network), None)
```

Reply:
I think I agree but it's out of scope of my PR here.

Reply:
Got it, makes sense